### PR TITLE
Remove Tensorflow

### PR DESCRIPTION
**Pull request by apchytr**
**Context:** `Tensorflow` has long been a part of Mr Mustard. However, in recent years it has caused a number of issues/limitations. Given this fact and that the greater software community has been generally moving away from `tensorflow` in favour of libraries such as `JAX` gives us enough of an incentive to move away from having `tensorflow` as a dependency. Here, we deprecate the `tensorflow` backend to instead favour `JAX`. 

**Description of the Change:** Removed all references to `tensorflow`. Changed some references to `tensorflow` to instead make use of `JAX` wherever appropriate. Removed `assign`, `new_variable`, `new_constant`, `from_backend`, `is_trainable`, `DefaultEuclideanOptimizer` from the math backend. `update_fn` in `Variable` is now a string.

**Benefits:** No more `tensorflow` issues / limitations. A more lightweight Mr Mustard experience. 

**Possible Drawbacks:** Some gradients still need to be ported to `JAX`

### CODE REVIEW

**Review by timmysilv — COMMENTED** [review]
very exciting

some changes here seem unrelated, eg. passing functions in as strings instead of references. did jax have some problems that needed this? Are the removed functions from backends ones that were only needed when using tensorflow?

Also, I still see a few references to tensorflow in the code:
```bash
$ grep -irl tensorflow mrmustard --exclude-dir __pycache__
mrmustard/__init__.py
mrmustard/math/lattice/strategies/compactFock/inputValidation.py
mrmustard/math/parameters.py
mrmustard/README.md
```

File: doc/code/training/optimizer.rst [hunk 1]
```diff
@@ -1,7 +1,7 @@
 Optimizer
 =========
 
-.. currentmodule:: mrmustard.training.optimizer
+.. currentmodule:: mrmustard.training.optimizer_jax
```
Comment (heltluke) [inline doc/code/training/optimizer.rst]: This feels at odds with renaming `OptimizerJax` as `Optimizer`.

**Comment by apchytr** [comment]
> some changes here seem unrelated, eg. passing functions in as strings instead of references. did jax have some problems that needed this? Are the removed functions from backends ones that were only needed when using tensorflow?

Probably not great but I lumped in removing a bunch of legacy stuff that was only there to support tensorflow's optimizations. This included some changes to `Variable` and `Constant` like changing `update_fn` to be a string instead of a callable. This is so that I could remove the implementations of the various custom update functions for Riemannian optimizations which were only used by `tf`. If this PR is too large for reviewers I can see that it's split up just lmk!

> Also, I still see a few references to tensorflow in the code:
> 
> ```shell
> $ grep -irl tensorflow mrmustard --exclude-dir __pycache__
> mrmustard/__init__.py
> mrmustard/math/lattice/strategies/compactFock/inputValidation.py
> mrmustard/math/parameters.py
> mrmustard/README.md
> ```
 Thanks for pointing these out! Updated 👍

File: mrmustard/lab/states/gaussian_state.py [hunk 2]
```diff
@@ -84,7 +83,7 @@ def __init__(
                 value=symplectic,
                 name="symplectic",
                 bounds=(None, None),
-                update_fn=update_symplectic,
+                update_fn="update_symplectic",
```
Comment (heltluke) [inline mrmustard/lab/states/gaussian_state.py:86]: What is this change for?

File: mrmustard/lab/utils.py [hunk 3]
```diff
@@ -18,19 +18,19 @@
...
+from mrmustard.math.parameters import Constant, Variable
 
 
 def make_parameter(
     is_trainable: bool,
     value: Any,
     name: str,
     bounds: tuple[float | None, float | None],
-    update_fn: Callable = update_euclidean,
+    update_fn: str = "update_euclidean",
```
Comment (heltluke) [inline mrmustard/lab/utils.py]: Is this more truly a `Literal` than a `str`? That is, is there a finite set of `update_fn`s?

**Comment by apchytr** [comment]
> This included some changes to `Variable` and `Constant` like changing `update_fn` to be a string instead of a callable. 

I should add that we are already discussing ideas on how to update our API to better adhere to JAX and refactoring `Variable` and `Constant` is part of the plan

File: mrmustard/math/backend_jax.py [hunk 4]
```diff
@@ -181,24 +175,10 @@ def conj(self, array: jnp.ndarray) -> jnp.ndarray:
     def pow(self, x: jnp.ndarray, y: float) -> jnp.ndarray:
         return jnp.power(x, y)
 
-    def new_variable(
```
Comment (heltluke) [inline mrmustard/math/backend_jax.py:184]: Why is it that you're able to delete this and other methods here?

File: mrmustard/math/parameters.py [hunk 5]
```diff
@@ -118,55 +113,6 @@ def format_value(param: Constant | Variable) -> tuple[str, str]:
     return value_str, "scalar"
 
 
-def update_symplectic(grads_and_vars, symplectic_lr: float):
```
Comment (heltluke) [inline mrmustard/math/parameters.py:121]: Why are you able to delete these three functions?

File: mrmustard/physics/ansatz/polyexp_ansatz.py [hunk 6]
```diff
@@ -786,9 +786,7 @@ def _partial_eval(self, z: ArrayLike, indices: tuple[int, ...]) -> PolyExpAnsatz
         z_batch_idxs = tuple(range(self.batch_dims, self.batch_dims + len(z_batch_shape)))
         z = math.transpose(
             math.broadcast_to(z, self.batch_shape + z.shape),
-            z_batch_idxs
-            + ansatz_batch_idxs
-            + (len(z_batch_idxs + ansatz_batch_idxs),),  # tensorflow
+            z_batch_idxs + ansatz_batch_idxs + (len(z_batch_idxs + ansatz_batch_idxs),),
```
Comment (heltluke) [inline mrmustard/physics/ansatz/polyexp_ansatz.py:789]: This looks the same as before, but without the `# tensorflow` comment. Or did I miss something?

File: mrmustard/training/__init__.py [hunk 7]
```diff
@@ -29,41 +29,38 @@
...
-    from mrmustard.lab.gates import S2gate, BSgate
-    from mrmustard.lab.states import Vacuum
-    from mrmustard.lab.circuit import Circuit
+    from mrmustard.lab import BSgate, Circuit, S2gate, Vacuum
     from mrmustard.training import Optimizer
 
     r = np.arcsinh(1.0)
-    s2_0 = S2gate(r=r, phi=0.0, phi_trainable=True)[0, 1]
-    s2_1 = S2gate(r=r, phi=0.0, phi_trainable=True)[2, 3]
+    s2_0 = S2gate(modes=(0,1), r=r, phi=0.0, phi_trainable=True)
```
Comment (heltluke) [inline mrmustard/training/__init__.py]: Thanks for updating the example! Maybe a sign we should run `doctest` against it?

File: tests/test_math/test_compactFock.py [hunk 8]
```diff
@@ -65,7 +65,7 @@ def test_compactFock_1leftover():
     assert np.allclose(expected, G_leftover)
 
 
-@pytest.mark.requires_backend("tensorflow")
+@pytest.mark.requires_backend()  # TODO: implement gradient of hermite_renormalized_diagonal
```
Comment (heltluke) [inline tests/test_math/test_compactFock.py:68]: As part of this PR, or in a subsequent one?

File: tests/test_physics/test_gaussian_integrals.py [hunk 9]
```diff
@@ -48,7 +48,7 @@ def test_real_gaussian_integral():
             ],
         ],
     )
-    b = math.cast(math.arange(3), dtype=math.complex128)  # tensorflow does not support complex
+    b = math.cast(math.arange(3), dtype=math.complex128)
```
Comment (heltluke) [inline tests/test_physics/test_gaussian_integrals.py]: What was the comment meant to tell us?

File: tests/test_training/test_opt_lab.py
Comment (heltluke) [inline tests/test_training/test_opt_lab.py:1]: Are these tests part of `test_optimizer` now?

File: tests/test_training/test_riemannian_opt.py
Comment (heltluke) [inline tests/test_training/test_riemannian_opt.py:1]: Are these tests reproduced in another file?

File: tests/test_utils/test_serialize.py [hunk 10]
```diff
@@ -140,28 +139,6 @@ def test_overlap_forbidden(self):
...
-    @pytest.mark.requires_backend("tensorflow")
-    def test_backend_change_error(self, monkeypatch):
-        """Test that data must be deserialized with the same backend."""
-        x = math.astensor([1.1, 2.2])
-        path = save(DummyOneNP, name="myname", arrays={"array": x})
-        # can be thought of as restarting python and not changing to tensorflow
-        monkeypatch.setattr("mrmustard.math._backend._name", "numpy")
-        with pytest.raises(
-            TypeError,
-            match="Data serialized with tensorflow backend, cannot deserialize to the currently active numpy backend",
```
Comment (heltluke) [inline tests/test_utils/test_serialize.py:160]: Does this kind of error not occur between `numpy` and `jax`?

File: mrmustard/training/optimizer.py
Comment (heltluke) [inline mrmustard/training/optimizer.py:1]: I'll admit that I did not review this file in great detail, but it does seem like the `Optimizer` class has changed a fair bit. Perhaps you could summarize the changes a little, or draw reviewers' attention to anything of note?

File: doc/code/training/optimizer.rst (hunk 1 above)
Comment (apchytr) [reply doc/code/training/optimizer.rst]: Just updated!

File: mrmustard/lab/states/gaussian_state.py (hunk 2 above)
Comment (apchytr) [reply mrmustard/lab/states/gaussian_state.py:86]: See [this comment](https://github.com/XanaduAI/MrMustard/pull/633#issuecomment-3113717098). In summary, this was needed to remove some legacy tf stuff. The functions themselves were not used in JAX

File: mrmustard/math/backend_jax.py (hunk 4 above)
Comment (apchytr) [reply mrmustard/math/backend_jax.py:184]: These were legacy tensorflow methods not required for JAX 😁

File: mrmustard/math/parameters.py (hunk 5 above)
Comment (apchytr) [reply mrmustard/math/parameters.py:121]: Ditto from the above

File: mrmustard/physics/ansatz/polyexp_ansatz.py (hunk 6 above)
Comment (apchytr) [reply mrmustard/physics/ansatz/polyexp_ansatz.py:789]: Wasn't sure what change I originally had in mind so removed it

File: mrmustard/training/__init__.py (hunk 7 above)
Comment (apchytr) [reply mrmustard/training/__init__.py]: Yep! We definitely should be running `doctest`

File: tests/test_math/test_compactFock.py (hunk 8 above)
Comment (apchytr) [reply tests/test_math/test_compactFock.py:68]: I'm thinking subsequent since this PR is already substantial

File: tests/test_physics/test_gaussian_integrals.py (hunk 9 above)
Comment (apchytr) [reply tests/test_physics/test_gaussian_integrals.py]: ~~No idea 😅~~ I believe it was just in reference to having to use `math.cast` instead of the dtype argument of `arange`. Updated now

File: tests/test_training/test_opt_lab.py
Comment (apchytr) [reply tests/test_training/test_opt_lab.py:1]: Was ported as part of [#615](https://github.com/XanaduAI/MrMustard/pull/615)

File: tests/test_training/test_riemannian_opt.py
Comment (apchytr) [reply tests/test_training/test_riemannian_opt.py:1]: Not identically. My logic was that the test_optimizer.py tests were already making sure the update methods worked. It might be good to just port these directly as well?

File: tests/test_utils/test_serialize.py (hunk 10 above)
Comment (apchytr) [reply tests/test_utils/test_serialize.py:160]: Not sure. Our serialization isn't used frequently so hard to tell. @timmysilv was the one who implemented it so maybe he knows?

File: mrmustard/training/optimizer.py
Comment (apchytr) [reply mrmustard/training/optimizer.py:1]: Nothing in this file actually changed! This was just a rename of `OptimizerJax` to `Optimizer`

File: mrmustard/math/backend_jax.py (hunk 4 above)
Comment (heltluke) [reply mrmustard/math/backend_jax.py:184]: I don't _quite_ understand why it's in `backend_jax` if it was only needed for `tensorflow`. It's something like a "default" implementation is good enough, without looking up `numpy` vs `jax` vs `tensorflow`?

File: tests/test_math/test_compactFock.py (hunk 8 above)
Comment (heltluke) [reply tests/test_math/test_compactFock.py:68]: Could I ask you to make a Shortcut Story, if you haven't already, so that it's not forgotten?

File: tests/test_training/test_riemannian_opt.py
Comment (heltluke) [reply tests/test_training/test_riemannian_opt.py:1]: It might be, yeah. It's not obvious to me whether the tests there are sufficient or not.

File: tests/test_utils/test_serialize.py (hunk 10 above)
Comment (heltluke) [reply tests/test_utils/test_serialize.py:160]: If it can, this test should simply be updated, not removed, right?

File: mrmustard/training/optimizer.py
Comment (heltluke) [reply mrmustard/training/optimizer.py:1]: Oh, and GitHub has done a terrible job of noticing that. OK!

File: mrmustard/math/backend_jax.py (hunk 4 above)
Comment (apchytr) [reply mrmustard/math/backend_jax.py:184]: Yeah, just the nature of the `math` backend. If we have some backend specific method then the other ones have to have these dummy default implementations

File: tests/test_utils/test_serialize.py (hunk 10 above)
Comment (timmysilv) [reply tests/test_utils/test_serialize.py:160]: it should be identical if you just replace tensorflow with jax. The library [saves the backend](https://github.com/XanaduAI/MrMustard/blob/0a947b454ad095426e66566122ed2af2700c0ebb/mrmustard/utils/serialize.py#L71) when serializing an object, then [ensures you have the same backend active](https://github.com/XanaduAI/MrMustard/blob/0a947b454ad095426e66566122ed2af2700c0ebb/mrmustard/utils/serialize.py#L116) when deserializing that object.

File: mrmustard/training/__init__.py (hunk 7 above)
Comment (apchytr) [reply mrmustard/training/__init__.py]: Updated and also added a task to https://app.shortcut.com/xanaduai/story/94960/documentation (outside of the scope of this PR)

File: mrmustard/lab/utils.py [hunk 11]
```diff
@@ -18,19 +18,21 @@
...
 
 def make_parameter(
     is_trainable: bool,
     value: Any,
     name: str,
     bounds: tuple[float | None, float | None],
-    update_fn: Callable = update_euclidean,
+    update_fn: Literal[
+        "update_euclidean", "update_orthogonal", "update_symplectic", "update_unitary"
+    ] = "update_euclidean",
```
Comment (timmysilv) [inline mrmustard/lab/utils.py:35]: if this type will ever change, we should probably define it once in `math`

**Review by timmysilv — DISMISSED** [review]

File: mrmustard/lab/utils.py (hunk 11 above)
Comment (apchytr) [reply mrmustard/lab/utils.py:35]: The more likely option is that the need for `update_fn` will go away. Imo `Variable` itself needs to be rethought for JAX

File: tests/test_math/test_compactFock.py (hunk 8 above)
Comment (apchytr) [reply tests/test_math/test_compactFock.py:68]: https://app.shortcut.com/xanaduai/story/95226/cleanup-lattice-methods

**Review by timmysilv — DISMISSED** [review]

File: tests/test_utils/test_serialize.py [hunk 12]
```diff
@@ -140,24 +139,16 @@ def test_overlap_forbidden(self):
         ):
             save(Dummy, arrays={"val": [1]}, val=2)
 
-    @pytest.mark.requires_backend("tensorflow")
-    def test_tensorflow_support(self):
```
Comment (heltluke) [inline tests/test_utils/test_serialize.py:144]: I suppose you could port this for `jax` as well?

**Review by heltluke — DISMISSED** [review]

**Review by timmysilv — APPROVED** [review]

**Review by ziofil — APPROVED** [review]
yay!
